components/script/dom/datatransfer.rs

```diff
@@ -0,0 +1,191 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+
+use dom_struct::dom_struct;
+use js::rust::{HandleObject, MutableHandleValue};
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::DataTransferBinding::DataTransferMethods;
+use crate::dom::bindings::frozenarray::CachedFrozenArray;
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::datatransferitemlist::DataTransferItemList;
+use crate::dom::element::Element;
+use crate::dom::filelist::FileList;
+use crate::dom::htmlimageelement::HTMLImageElement;
+use crate::dom::window::Window;
+use crate::script_runtime::{CanGc, JSContext};
+
+const VALID_DROP_EFFECTS: [&str; 4] = ["none", "copy", "link", "move"];
+const VALID_EFFECTS_ALLOWED: [&str; 9] = [
+    "none",
+    "copy",
+    "copyLink",
+    "copyMove",
+    "link",
+    "linkMove",
+    "move",
+    "all",
+    "uninitialized",
+];
+
+// An image that can be shown when a drag event occurs
+#[derive(JSTraceable, MallocSizeOf)]
+#[crown::unrooted_must_root_lint::must_root]
+struct DragBitmap {
+    image: Dom<HTMLImageElement>,
```

Comment:
Again here, instead of storing the dome element, you should only store the image data, which would be probably this: 

https://github.com/servo/servo/blob/672b37dd9c4b27fb9892ce7148d163444bf10de0/components/script/dom/htmlimageelement.rs#L160